Woodoku/tests/entity/test_woodoku_board.py

```diff
@@ -0,0 +1,163 @@
+        first_shape and first_location are chosen so that sec_shape will be overlapped with first_shape when
+        adding it at sec_location
+        """
+        board = WoodokuBoard()
+        board.add_shape(first_shape, *first_position)
+        assert not board.can_add_shape_at_location(sec_shape, *sec_position)
+
+    @pytest.mark.parametrize('shape', [
+        l_shape,
+        horizontal_bar_shape,
+        gun_shape
+    ])
+    def test_can_add_shape_to_board_on_empty_board(self, shape: WoodokuShape) -> None:
+        board = WoodokuBoard()
+        assert board.can_add_shape_to_board(shape)
+
+    def test_can_add_shape_to_board_on_conflicted_board(self) -> None:
+        """
+        first_shape and first_location are chosen so that sec_shape will be overlapped with first_shape when
+        adding it at sec_location
+        """
+        board = WoodokuBoard()
+        for i in range(9):
+            assert board.can_add_shape_to_board(self.horizontal_bar_shape)
+        assert not board.can_add_shape_to_board(self.horizontal_bar_shape)
+
+    def test_find_groups_a_row(self) -> None:
+        """
+        A first row should be the only group on the board
+        """
+        board = WoodokuBoard()
+        board.add_shape(self.horizontal_bar_shape, 0, 0)
+        board.add_shape(self.gun_shape, 0, 5)
+        board.add_shape(self.vertical_two_block, 0, 8)
+        info, group_blocks = board._WoodokuBoard__find_groups()
+        assert info == {"row": 1, "column": 0, "box": 0}
+        assert group_blocks == set([(0, col) for col in range(9)])
+
+    def test_find_groups_almost_a_col(self) -> None:
+        """
+        A first column should be the only group on the board
+        """
+        board = WoodokuBoard()
+        board.add_shape(self.gun_shape, 0, 1)
+        board.add_shape(self.vertical_two_block, 2, 1)
+        board.add_shape(self.gun_shape, 4, 1)
+        board.add_shape(self.vertical_two_block, 6, 1)
+        info, group_blocks = board._WoodokuBoard__find_groups()
+        assert info == {"row": 0, "column": 0, "box": 0}
+
+    def test_find_groups_a_col(self) -> None:
+        board = WoodokuBoard()
+        board.add_shape(self.gun_shape, 0, 1)
+        board.add_shape(self.vertical_two_block, 2, 1)
+        board.add_shape(self.gun_shape, 4, 1)
+        board.add_shape(self.vertical_two_block, 6, 1)
+        board.can_add_shape_at_location(self.l_shape, 7, 0)
+        info, group_blocks = board._WoodokuBoard__find_groups()
+        assert info == {"row": 0, "column": 1, "box": 0}
+        assert group_blocks == set([(row, 1) for row in range(9)])
+
+    def test_find_groups_a_three_by_three_block(self) -> None:
+        """
+        A 3X3 square at the right corner should be the only group on the board
+        """
+        board = WoodokuBoard()
+        board.add_shape(self.gun_shape, 0, 0)
+        board.add_shape(self.horizontal_bar_shape, 1, 1)
+        board.add_shape(self.gun_shape, 2, 0)
```

Comment:
Ditto, `add_shape` shall remove the 3x3 block so `_WoodokuBoard__find_groups` shows nothing.